### ISSUE TITLE
Xtext validation not removing markers in Eclipse?

### ISSUE THREAD

**Issue by szschaler** [eclipse/xtext-eclipse#462]
Hi,

There seems to be some inconsistency in how Xtext handles Eclipse markers (or possibly its own annotations). I had originally asked a question about this in the [forum](https://www.eclipse.org/forums/index.php/m/1776927/). After some discussion, @cdietrich suggested this may actually be a bug, so I am reporting this here.

Specifically, I have an expensive validation that only needs to run if a certain flag is set in my model. I set up the validation as follows (DSML elements faked for the purposes of this post):

```
class MyDSLValidator ....

  @Check(EXPENSIVE)
  def void myExpensiveCheck (Model m) {
    if (m.flag) {
      // perform expensive validation
      error ("Expensive check went wrong", ...)
    }
```

In order to flag up to the user that there may be an error, but we haven't checked this yet, I also add a warning marker to the flag through a FAST check:

```
  @Check
  def void warnAboutMissingCheck (Model m) {
    if (m.flag && !checkMode.shouldCheck(CheckType.EXPENSIVE)) {
      warning("Please run validate for complete checking.", ...)
    }
  }
}
```

As expected, when I create a model in the Xtext editor and enable the flag (by providing a suitable keyword), the warning marker is created and shown in the editor. When I run Validate, I also get the error marker (assuming the model had the problem I'm checking for), but the warning marker is also still there. I have debugged through my code; `warnAboutMissingCheck` is run when I run Validate, but the check works and the warning is not added in that case.

It seems there is no clean Xtext way to remove annotations / markers once they've been created.

@cdietrich had the following thoughts on the [forum](https://www.eclipse.org/forums/index.php/m/1776927/):

> for me this could be a bug .
>org.eclipse.xtext.ui.editor.validation.MarkerIssueProcessor.processIssues(List<Issue>, IProgressMonitor)
>seems to add stuff only but not remove existing ones.
>
>org.eclipse.xtext.ui.editor.validation.AddMarkersOperation.execute(IProgressMonitor) deletes the markers.
>org.eclipse.ui.texteditor.ResourceMarkerAnnotationModel.batchedUpdate(IMarkerDelta[])
>treats thst in the editor.
>
>but thats marker annotations only.
>this leaves the xtext annotations unattended
>
>these are handle by org.eclipse.xtext.ui.editor.validation.AnnotationIssueProcessor.
>
>imho the validationjob should use a hybrid of AnnotationIssueProcessor and MarkerIssueProcessor (in ValidateActionHandler)

Many thanks,

Steffen

**Comment by cdietrich** [comment]
@szarnekow do you have any idea how to cope with this problem

**Comment by szarnekow** [comment]
Expensive validation markers are not a good citizen. I recommend to use a custom IXtextEditorCallback to deal with these markers explicitly.

**Comment by cdietrich** [comment]
@szarnekow the other way round. the editor markers are the problem (XtextAnnotation). they are not removed

**Comment by szarnekow** [comment]
My bad, I meant an editor callback to handle the annotations. See org.eclipse.xtext.ui.editor.validation.ValidatingEditorCallback as an inspiration.

**Comment by szschaler** [comment]
Thanks. I have looked at the code of `org.eclipse.xtext.ui.editor.validation.ValidatingEditorCallback`, but am not sure I understand what your proposal is. That code seems to essentially schedule the non-expensive validations at editor creation and editor saving. 

My problem is that the non-expensive markers aren't removed when I run an expensive validation explicitly.

Could you be a little bit more explicit about what you were suggesting I should do?

**Comment by szarnekow** [comment]
I was thinking in the direction of a custom `AnnotationIssueProcessor`. That one should handle expensive annotations even if the current run was not for expensive checks.
It would have to be registered on the `IXtextDocument`. One way to do that is via the `IXtextEditorCallback` or you could customize the `XtextDocumentProvider.registerAnnotationInfoProcessor(ElementInfo)`

**Comment by szschaler** [comment]
I see. Many thanks. 

However, I don't think this would solve my problem. If I understand, you're proposing a solution that would run the expensive checks even when they're not meant to be run. I am happy with **when** my expensive check is run. The reason they aren't run at edit or save time is that they can potentially become really expensive (we're potentially talking exponential here).

What I am not happy with is what happens to other annotations when the expensive checks are run. In particular, I have a cheap warning annotation that is generated during normal cheap runs of the validation. This should disappear when the expensive check is run. My code ensures it only generates the cheap warning when we're not running in expensive mode. Still, the annotation marker appears in the editor.

I'm not sure I understand how running the expensive checks earlier would fix this problem.

**Comment by ArneDeutsch** [comment]
I have had a look about this issue. Currently I do know the reason but not yet the solution.

If a validation rule from the expensive validation job is run a resource marker is created. From this resource marker ONE annotation is created in the the editor. Removing the marker will lead to removal of the annotation, leading the remoal of the error from the edtior, On the other hand when the editor itself runs the validation TWO annotations are added. One from the editor (because the resource might not yet have been changed) and one from the resource after the change has been saved. If now the resource marker is deleted  one ONE of the TWO annotations gets removed. Althought the error disappears from the error view it is still visible in the editor.

I think it is this inconsistency between errors created from the editor and the ones created from the validation that leads to issues such as this one.

The above can be validated quite easaly by createing an example grammer with two rules (one cheap, one expensive) creating an error marker all the times ... and then removing the markers with the context menu in the problems view. The annotation for the cheap validation will remain while the expensive one remains.

**Comment by szschaler** [comment]
Excellent, many thanks for this fix. Is there an easy way for me to try this out or is it better if I wait until that release?

**Comment by kthoms** [comment]
The fix is just provided, but not approved and merged yet. When this is done you can consume Xtext from the nightly build repo http://download.eclipse.org/modeling/tmf/xtext/updates/composite/latest/

**Comment by szschaler** [comment]
Thanks.

**Comment by cdietrich** [comment]
@kthoms @ArneDeutsch is there anything left for this issue?

**Comment by ArneDeutsch** [comment]
From my point of view this is done. As @kthoms has merged the fix I think he has also reviewed. I would say we can close this ticket as done.

**Comment by cdietrich** [comment]
Fixed in Xtext 2.14
